src/raftstore/store/snap.rs

```diff
@@ -61,223 +108,519 @@ impl Display for SnapKey {
     }
 }
 
-/// Name prefix for the self-generated snapshot file.
-const SNAP_GEN_PREFIX: &'static str = "gen";
-/// Name prefix for the received snapshot file.
-const SNAP_REV_PREFIX: &'static str = "rev";
+/// `SnapshotGenerator` is a trait that generates snapshot from `RocksDB` snapshot,
+/// the generated snapshot is used to replicate local db data to other remote raftstores.
+pub trait SnapshotGenerator {
+    fn generate(&mut self,
+                snap: &DbSnapshot,
+                region: &Region,
+                snap_data: &mut RaftSnapshotData)
+                -> raft::Result<()>;
+}
 
-/// A structure represents the snapshot file.
-///
-/// All changes to the file will be written to `tmp_file` first, and use
-/// `save` method to make them persistent. When saving a crc32 checksum
-/// will be appended to the file end automatically.
-pub struct SnapFile {
-    file: PathBuf,
-    digest: Digest,
-    size_track: Arc<RwLock<u64>>,
-    // File is the file obj represent the tmpfile, string is the actual path to
-    // tmpfile.
-    tmp_file: Option<(File, String)>,
+/// `Snapshot` is a trait for snapshot.
+/// It's used in these scenarios:
+///   1. read local snapshot and then replicate it to remote raftstores
+///   2. receive snapshot from remote raftstore and write it to local storage
+///   3. snapshot gc
```

Comment:
The snapshots which are generated locally or received from remote raftstores will be GCed regularly in raftstore.